### PR TITLE
feat: execute sale listing rewards program

### LINKED ISSUE #13: Execute Sale
### Goal
Complete the sale of an NFT when a listing matches an offer.
Payout spl token to both the buyer and the seller based on multiple of the SOL sale price.
The split of spl token to the buyer and the seller is configured on the reward center listing reward rules

### Requirements
- Mark the purchased_at on the listing and offer using clock current time.
- CPI call to auctioneer_execute_sale
- Listing reward rules tracks seller_reward_payout_basis_points. The rest of the payout goes to the buyer.
- Listing rewards tracks payout_divider which is used to determine the amount of SPL token to payout compared to the lamports. For example, if the payout_divider is 2 and the sale was for 2,000,000,000 lamports the user will get 1,000,000,000 SPL token (or half the amount of the sale)

### PR DESCRIPTION
**Pull request by sunguru98**
Added Execute sale instruction which completes the sale and transfers rewards

Closes #13

### CODE REVIEW

File: listing-rewards/program/src/execute_sale/mod.rs [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+        escrow_payment_bump,
+        free_trade_state_bump,
+        program_as_signer_bump,
+        price,
+        token_size,
+        ..
+    }: ExecuteSaleParams,
+) -> Result<()> {
+    let seller_listing = &mut ctx.accounts.listing;
+    let buyer_offer = &mut ctx.accounts.offer;
```
Comment (kespinola) [inline listing-rewards/program/src/execute_sale/mod.rs:294]: With auctioneer its the responsibility of the third party program to enforce price matching which means we need an assertion that the offer amount is the same as the listing amount.

```rust
assert_eq!(offer.price, listing.price, ListingRewardsError::PriceMismatch)?;
```

File: listing-rewards/program/src/execute_sale/mod.rs [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+    };
+
+    invoke_signed(
+        &execute_sale_ix,
+        &execute_sale_ctx_accounts.to_account_infos(),
+        &reward_center_signer_seeds,
+    )?;
+
+    let (seller_payout, buyer_payout) = reward_center.payouts(price)?;
+    // Buyer transfer
```
Comment (kespinola) [inline listing-rewards/program/src/execute_sale/mod.rs:377]: Should check if the reward treasury has enough to cover the payouts. If it doesn't then complete the sale without doing the token payout so basically exit doing the execute_sale invoke_signed

**Review by kespinola — CHANGES_REQUESTED** [review]

File: listing-rewards/program/src/execute_sale/mod.rs (hunk 1 above)
Comment (sunguru98) [reply listing-rewards/program/src/execute_sale/mod.rs:294]: Ahh, good catch. I totally missed that.

File: listing-rewards/program/src/execute_sale/mod.rs (hunk 2 above)
Comment (sunguru98) [reply listing-rewards/program/src/execute_sale/mod.rs:377]: Another good catch.

File: listing-rewards/program/src/execute_sale/mod.rs [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+                    to: ctx.accounts.buyer_reward_token_account.to_account_info(),
+                },
+                &reward_center_signer_seeds,
+            ),
+            buyer_payout,
+        )?;
+    }
+
+    // Seller transfer
+    ctx.accounts.reward_center_reward_token_account.reload()?;
```
Comment (kespinola) [inline listing-rewards/program/src/execute_sale/mod.rs:404]: Nice. Great job on the reload!

**Review by kespinola — APPROVED** [review]

**Pull request merged**